### PR TITLE
The future of Keycloak Operator CRs

### PR DESCRIPTION
**Pull request by vmuzikar**
**This has to be merged after https://github.com/keycloak/keycloak-k8s-resources/pull/11**

### CODE REVIEW

File: blog/2022/operator-crs.adoc
Comment (DGuhr) [inline blog/2022/operator-crs.adoc:22]:
```suggestion
CRs will be added once the new store has support for it, which is currently expected in Keycloak 21 (please note this is
```

Comment (vmuzikar) [reply blog/2022/operator-crs.adoc:22]: Good catch, thanks!

**Review by DGuhr — APPROVED** [review]
LGTM

**Review by pedroigor — APPROVED** [review]

File: blog/2022/operator-crs.adoc
Comment (abstractj) [inline blog/2022/operator-crs.adoc:18]:
```suggestion
reliability, and predictability to the whole solution.
```

**Review by abstractj — APPROVED** [review]
@vmuzikar LGTM, added a small suggestion.

File: blog/2022/operator-crs.adoc
Comment (vmuzikar) [reply blog/2022/operator-crs.adoc:18]: 👍 Thanks!

**Comment by vmuzikar** [comment]
Just a small note. Before merging, I would like a review from a Storage team member to double check the statements are true.

**Comment by abstractj** [comment]
cc @keycloak/storage-x

**Review by sguilhen — APPROVED** [review]

**Comment by sguilhen** [comment]
I think this is accurate but would wait for confirmation from @hmlnarik

**Comment by vmuzikar** [comment]
> I think this is accurate but would wait for confirmation from @hmlnarik

@hmlnarik Ping? :)

File: blog/2022/operator-crs.adoc
Comment (hmlnarik) [inline blog/2022/operator-crs.adoc:20]: It is not only the new storage which lacks immutability. Non-exhaustive set of examples of inter-area dependencies follows:

- When a new realm is added (even via a CR), master realm has to be updated with new role and client
- Realm default client scopes needs to be cleaned on client scope removal
- When a role is removed, respective role mapping in groups and users is updated.
- When a client is removed (even via CR), respective client roles are removed as well.

These are not mandated by storage but by the Keycloak design.

As mentioned in a meeting earlier, immutability in the storage is one of features which can be implemented in the new storage. However, without changing the design and implementation, storage immutability cannot be enabled while keeping Keycloak operable.

File: blog/2022/operator-crs.adoc
Comment (hmlnarik) [inline blog/2022/operator-crs.adoc:23]: Seems some of CRs should be CRDs?

```suggestion
However, since the new storage is still work in progress and currently is lacking some features like the options for immutability that
are crucial for the new CRs, the new Operator is currently missing those CRs for managing Keycloak resources. The missing
CRDs will be added once the new store has support for it, which is currently expected in Keycloak 21 (please note this is
subject to change).
```

File: blog/2022/operator-crs.adoc
Comment (hmlnarik) [inline blog/2022/operator-crs.adoc:19]:
```suggestion
One of the most common concerns around the new Operator is the current lack of the CRDs for managing Keycloak resources, such as realm, users and clients, in a cloud-native way. One of the key aspects of the new Operator will be redesign
of managing these Keycloak resources via CRs and git-ops. This new approach will leverage
the https://www.keycloak.org/2022/07/storage-map.html[new storage architecture] and future immutability options,
making the CRs the declarative single source of truth. In comparison to the legacy Operator, this will bring high robustness,
reliability, and predictability to the whole solution.
```

Immutability is far more wired-in than in the storage, see comment below.

File: blog/2022/operator-crs.adoc
Comment (hmlnarik) [inline blog/2022/operator-crs.adoc:6]:
```suggestion
A while back, we have https://www.keycloak.org/2021/10/keycloak-x-update.html#_operator_and_containers[announced] a new
```

File: blog/2022/operator-crs.adoc
Comment (vmuzikar) [reply blog/2022/operator-crs.adoc]: Thanks for the review! Good point. Rephrased that section a bit. Please let me know if it works for you.

File: blog/2022/operator-crs.adoc
Comment (hmlnarik) [inline blog/2022/operator-crs.adoc:26]: This is a big topic shift between paragraphs: operator -> new storage :)

```suggestion
Before we would consider operator ready for leveraging CRs, we expect completing several features:

* Read-only administration REST API
* Read-only view in the console and other interfaces
* File store (expected in Keycloak 20).

All of this is critical to proper CRs implementation,
hence the new Operator is currently missing the CRDs for managing Keycloak resources. The missing CRDs will be added once
Keycloak has the necessary support for it, which is currently expected in Keycloak 21.
```

File: blog/2022/operator-crs.adoc
Comment (hmlnarik) [inline blog/2022/operator-crs.adoc:28]: Repository referred by this URL is empty (?)

Also would be good to start this paragraph with explanation of what workarounds could be expected so that the reader would know prior to following the link.

File: blog/2022/operator-crs.adoc
Comment (vmuzikar) [reply blog/2022/operator-crs.adoc]: Thank you for the feedback! I too an inspiration from your suggestion and updated the PR. Please let me know if it works for you.

File: blog/2022/operator-crs.adoc
Comment (hmlnarik) [reply blog/2022/operator-crs.adoc:32]: Then I would suggest referring directly to README file and stating it:
```suggestion
We have prepared a few options to alleviate the situation with missing CRDs
in https://github.com/keycloak/keycloak-k8s-resources/tree/operators-tandem[this repository].
If you would like to try the new operator out, please see those options described in
in https://github.com/keycloak/keycloak-k8s-resources/blob/operators-tandem/README.md[README] file.
```
WDYT?

Comment (vmuzikar) [reply blog/2022/operator-crs.adoc]: Thanks, your wording seems better to me. Updated the PR.

The README is in the root of the branch, so when a user clicks the link to the branch itself, it opens the readme for them too. Therefore I wouldn't put a second link just to the README there.

**Review by ahus1 — COMMENTED** [review]
As I've been added as a reviewer, I'm commenting here. 

With this text, a reader might assume that once KC21 arrived, the CRDs will work as before, that is that all data about a realm can be stored inside a CR with all its attributes, clients etc.

IMHO we're not there yet to make such promises about a future design. With the upcoming file store, it might be more suitable to have the CR point to a Git reference with the data, or some other way where the data for the file store is passed as an opaque entity to the file store. 

The solution depends on a user's need, and a CR/CRD is only means to an end. Looking at users' scenarios and the capabilities of the file store, they might look different in the future. 

If you think that are reader won't have the expectation "once KC21 arrives, the CRDs will work as before", or if you decide that CR/CRDs need to work as before, then please disregard my comment.

**Comment by vmuzikar** [comment]
@ahus1 Thanks for the review! From my perspective it is clear now that we want to provide CRs for KC business objects, and that we need to provide a similar functionality to the old Operator – i.e. the capability to manage Realms, Clients, ... This is what the users and customers expect. Now, whether it will be really strictly typed CRDs, ConfigMap or some "free-form" CRDs, it's IMHO an implementation detail, as long as it is possible to manage the business objects from K8s environment. And in the blog post we're not stating the CRDs will look the same as with the old Operator.

Does it make sense?

**Review by ahus1 — APPROVED** [review]
@vmuzikar - thanks for stating your understanding that we are still open for different designs after this post. Approving it.

**Comment by vmuzikar** [comment]
@hmlnarik Can you please confirm the recent changes look good to you?

**Review by hmlnarik — APPROVED** [review]

File: blog/2022/operator-crs.adoc
Comment (stianst) [inline blog/2022/operator-crs.adoc:14]:
```suggestion
such as realm, users and clients, in a Kube-native way. One of the key aspects of the new Operator will be redesign
```

File: blog/2022/operator-crs.adoc
Comment (stianst) [inline blog/2022/operator-crs.adoc:24]: Admin UI shouldn't be fully read-only. It should only be read-only for immutable resources, while there should still be possibility to manage mutable resources.

Comment (vmuzikar) [reply blog/2022/operator-crs.adoc]: @stianst Thanks for the review! I'd try to loosen the statement about read-only to not make any commitments there. Even though initially we could really have e.g. Admin Console read-only as a whole.

**Review by stianst — APPROVED** [review]

**Pull request merged**
